src/mpr/parseMPR.ts

```diff
@@ -92,3 +85,251 @@ export class ParseHeader {
+    this.activeMaterialMass = buffer.readFloat32();
+    this.atX = buffer.readFloat32();
+    this.molecularWeight = buffer.readFloat32();
+    this.atomicWeight = buffer.readFloat32();
+    this.acquisitionStart = buffer.readFloat32();
+    this.eTransferred = buffer.readUint16(); // 3 bytes but only 2 are used
+    buffer.offset = zero + 0x11e;
+    this.electrodeMaterial = pascalString(buffer);
+    buffer.offset = zero + 0x1c0;
+    this.electrolyte = pascalString(buffer);
+    buffer.offset = zero + 0x211;
+    this.electrodeArea = buffer.readFloat32();
+    this.referenceElectrode = pascalString(buffer);
+    buffer.offset = zero + 0x24c;
+    this.characteristicMass = buffer.readFloat32();
+    this.batteryCapacity = buffer.readFloat32();
+    this.batteryCapacityUnit = buffer.readByte();
+    this.params = readParams(buffer, getParams(this.techniqueId), zero);
+  }
+}
+
+export function addData(
+  variable: Record<string, Array<number | string>>,
+  value: string | number,
+) {
+  if (!Object.prototype.hasOwnProperty.call(variable, 'data')) {
+    variable.data = [];
+  }
+  variable.data.push(value);
+}
+
+export function parseData(
+  buffer: IOBuffer,
+  header: ParseHeader,
+): Record<string, Record<string, Array<number | string> | string>> {
+  const zero = buffer.offset; // relative 0x0
+  const dataPoints = buffer.readUint32(); // Number of datapoints
+  const columns = buffer.readByte(); // Number of columns
+  const colIds = new Uint16Array(columns); // Id of the columns
+  const units = new Array<string>(columns); // Units for each column
+  for (let i = 0; i < columns; i++) {
+    const id = buffer.readUint16();
+    colIds[i] = id;
+    if (id in flagColumns) {
+      units[i] = 'flag';
+    } else if (id in dataColumns) {
+      units[i] = dataColumns[id][2];
+    }
+  }
+  //const data = new Array(dataPoints);
+
+  const variables: Record<
+    string,
+    Record<string, Array<number | string> | string>
+  > = {};
+  // Starts of datapoints vary between module versions
+  if (header.version <= 2) {
+    buffer.offset = zero + 0x195;
+  } else {
+    buffer.offset = zero + 0x196;
+  }
+
+  // Columns data gathering
+  for (let i = 0; i < dataPoints; i++) {
+    let flagByte = 256;
+    for (const id of colIds) {
+      if (id in flagColumns) {
+        if (flagByte === 256) flagByte = buffer.readByte();
+        const flag = flagColumns[id];
+        let twosComp = flag[0] & -flag[0];
+        let shift = -1;
+        while (twosComp > 0) {
+          twosComp >>= 1;
+          shift++;
+        }
+        if (!Object.keys(variables).includes(flag[1])) {
+          variables[flag[1]] = {};
+        }
+        const variab = variables[flag[1]];
+        addData(Object(variab), (flag[0] & flagByte) >> shift);
+        if (!Object.prototype.hasOwnProperty.call(variab, 'label')) {
+          variab.label = flag[1];
+        }
+        if (!Object.prototype.hasOwnProperty.call(variab, 'units')) {
+          variab.units = '';
```

Comment:
I'm wondring why you check hasOwnProperty and not simply `(if variab.units === undefined)`;

Reply:
Because I didn't know.